---
title: "Seurat Multimodal vignette"
output: github_document
fig_width: 6
fig_height: 4
---

# Introduction

Multimodal analysis is a great single cell genomics technique that allows the simultaneous measurement of multiple data types from the same cell.

- CITE-seq enables measurements of **transcriptomes** and **cell-surface proteins** from the same cell simultaneously

- 10xGenomics Multiome kit allows **cellular transcriptome** and **chromatin accessability** (scRNAseq + scATACseq)

Other modalities include : genetic perturbations, cellular methylomes, hashtag oligos from cellular hashing.

We will go through an introductory process but Seurat allows Weighted Nearest Neighbors (WNN) and more advanced techniques.

The data can be gathered from the [link](https://ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100866)

# Load libraries

```{r, message = FALSE, warning = FALSE, load-libraries}
library(tidyverse)
library(ggplot2)
library(Seurat)
library(patchwork)
```

# Load Dataset

```{r}
cbmc.rna <- as.sparse(read.csv(file = ".gitignore/GSE100866_CBMC_8k_13AB_10X-RNA_umi.csv.gz", sep = ",", header = TRUE, row.names = 1))
cbmc.rna <- CollapseSpeciesExpressionMatrix(cbmc.rna)

cbmc.adt <- as.sparse(read.csv(file = ".gitignore/GSE100866_CBMC_8k_13AB_10X-ADT_umi.csv.gz", sep = ",", header = TRUE, row.names = 1))
all.equal(colnames(cbmc.rna), colnames(cbmc.adt))
```
# Generate a Seurat Object

```{r, warning = FALSE}
cbmc <- CreateSeuratObject(counts = cbmc.rna)
# As default, seurat object stores as RNA
Assays(cbmc)

adt_assay <- CreateAssay5Object(counts = cbmc.adt)
cbmc[['ADT']] <- adt_assay

Assays(cbmc)

# Can be done in opposite way

#cbmc <- CreateSeuratObject(counts = cbmc.adt, assay = "ADT")
#rna_assay <- CreateAssay5Object(counts = cbmc.rna)
#cbmc[['RNA']] <- rna_assay

DefaultAssay(cbmc)
DefaultAssay(cbmc) <- "ADT"
DefaultAssay(cbmc) <- "RNA"
```
# Cluster cells on their scRNA-seq data 

```{r}
# Do the normal Seurat steps
cbmc <- NormalizeData(cbmc)
cbmc <- FindVariableFeatures(cbmc)
cbmc <- ScaleData(cbmc)
cbmc <- RunPCA(cbmc, verbose = FALSE)
cbmc <- FindNeighbors(cbmc, dims = 1:30)
cbmc <- FindClusters(cbmc, resolution = 0.8, verbose = FALSE)
cbmc <- RunUMAP(cbmc, dims = 1:30)
DimPlot(cbmc, label = TRUE)
```
But none of these take the non default assay into account

# Visualize multiple modalities side by side

```{r}
# Change the default assay and normalize and return
DefaultAssay(cbmc) <- "ADT"
cbmc <- NormalizeData(cbmc, normalization.method = "CLR", margin = 2)
DefaultAssay(cbmc) <- "RNA"

# This also works the same
# DefaultAssay(cbmc) is RNA
cbmc <- NormalizeData(cbmc, normalization.method = "CLR", margin = 2, assay = "ADT")
# DefaultAssay(cbmc) is still RNA

```

```{r, fig.width = 12, fig.height = 6}
DefaultAssay(cbmc) <- "ADT"
p1 <- FeaturePlot(cbmc, "CD19", cols = c("lightgrey", "darkgreen")) + ggtitle("CD19 protein")
DefaultAssay(cbmc) <- "RNA"
p2 <- FeaturePlot(cbmc, "CD19") + ggtitle("CD19 RNA")
p1 | p2
```
The umap was generated based on RNA analysis. We were able to visualize the ADT protein expression on the same map just after normalizing the data.

You can use specific assay using key values
```{r}
Key(cbmc[['RNA']])
Key(cbmc[['ADT']])
```
```{r, fig.width = 12, fig.height = 6}
p1 <- FeaturePlot(cbmc, "adt_CD19", cols = c("lightgrey", "darkgreen")) + ggtitle("CD19 protein")
p2 <- FeaturePlot(cbmc, "rna_CD19") + ggtitle("CD19 RNA")
p1 | p2
```
# Identify markers for scRNA-seq clusters

```{r}
VlnPlot(cbmc, "adt_CD19")
```
you can identify each markers using following ways

```{r}
adt_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "ADT")
rna_markers <- FindMarkers(cbmc, ident.1 = 6, assay = "RNA")

head(adt_markers)
head(rna_markers)

```

Additional Multimodal visualizations

```{r}
FeatureScatter(cbmc, feature1 = "adt_CD19", feature2 = "adt_CD3")
```

```{r}
FeatureScatter(cbmc, feature1 = "adt_CD3", feature2 = "rna_CD3E")
```

```{r}
FeatureScatter(cbmc, feature1 = "adt_CD4", feature2 = "adt_CD8")
```


