---
title: "R Notebook"
output: github_document
---

# Basics of Single cell RNA-seq

## Bulk RNA-Seq

-   estimate average expression level for each gene across a population of cells

-   use to characterize expression signatures between two groups or conditions ex) healthy / diseased

-   use to find and annotate new genes, gene isoforms, and other transcripts

## Single Cell RNA-Seq

-   allows comparison of trascriptomes in individual cells

-   assess transcriptional similarity and difference between cells

-   can find rare cell populations

-   find developmental relationships between heterogeneous cell states

# Key terms

## UMI (Unique Molecular Identifiers)

molecular tags that can be applied to detect and quantify the unique transcripts

## Features = genes

## Barcodes

Single cell sequencing experiments use short "DNA barcode" tags to identify reads that originate from the **same cell**

## Count Matrix / Feature-Barcode Matrix / Gene-Barcode Matrix

A Matrix of counts representing the number of unique observations of each feature within each cell barcode

## Doublets

**Two cells** are encapsulated into one reaction volume -\> should take them out

# Packages for single-cell RNA Seq Analysis

## R packages

-   SingleCellExperiment

-   Seurat

-   Monocle3

-   Scater

## Python

-   Scanpy

# Download Gene expression data from NCBI GEO

# Manipulate Gene experession data with dplyr

## Load Libraries

```{r, load-libraries}
library(dplyr)
library(tidyverse)
library(GEOquery)
```

## Read the data

```{r}
data <- read.csv("GSE183947_fpkm.csv")
```

## Get the metadata using GEO Query package

Geoquery practice = <https://www.bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html>

```{r}
gse <- getGEO(GEO = "GSE183947", GSEMatrix = TRUE)
gse
```

Get Phenotype data

```{r}
metadata <- pData(phenoData(gse[[1]]))
head(metadata)
```

## Use dplyr to process data

```{r}
metadata_subset <- metadata %>% 
  select(c(1, 10, 11, 17)) %>%
  rename(tissue = characteristics_ch1,
         metastasis = characteristics_ch1.1) %>%
  mutate(tissue = gsub("tissue: ", "", tissue),
         metastasis = gsub("metastasis: ", "", metastasis))
```

## join both datasets

```{r}
data_long <- data %>% rename(gene = X) %>%
  gather(key = 'samples', value = 'fpkm', -gene)
data_long <- data_long %>% left_join(metadata_subset, by = c("samples" = "description"))
head(data_long)
```

## Summarize the data

```{r}
data_long %>% filter(gene == "BRCA1" | gene == "BRCA2") %>%
  group_by(gene, tissue) %>%
  summarise(avg_fpkm = mean(fpkm),
            med_fpkm = median(fpkm)) %>%
  arrange(-avg_fpkm)
```
